src/mlmc/quantity.py

```diff
@@ -538,126 +538,113 @@ def samples(self, chunk_spec):
         """
-        Return reshaped mean
+        Calculates the overall estimates of the mean and the variance from the means and variances at each level
         """
-        return self.mean
+        self._mean = np.sum(self._l_means, axis=0)
+        self._var = np.sum(self._l_vars / self._n_samples[:, None], axis=0)
+
+        # Sums full of zeros leads to NaN
+        if np.isnan(self._mean).all():
+            self._mean = []
+            self._var = []
 
     @property
     def mean(self):
+        if self._mean is None:
+            self._calculate_mean_var()
         return self._reshape(self._mean)
 
     @property
     def var(self):
+        if self._var is None:
+            self._calculate_mean_var()
         return self._reshape(self._var)
 
     @property
     def l_means(self):
-        return self._reshape(self._l_means, levels=True)
+        return np.array([self._reshape(means) for means in self._l_means])
 
     @property
     def l_vars(self):
-        return self._reshape(self._l_vars, levels=True)
+        return np.array([self._reshape(vars) for vars in self._l_vars])
 
     @property
     def n_samples(self):
         return self._n_samples
 
-    def _reshape(self, data, levels=False):
-        if isinstance(self.qtype, qt.ArrayType):
-            reshape_shape = self.qtype._shape
-            size = self.qtype._qtype.size()
-            if isinstance(reshape_shape, int):
-                reshape_shape = [reshape_shape]
-            if levels:
-                return data.reshape((data.shape[0], *reshape_shape))
-            if size > 1:
-                return data.reshape(*reshape_shape, size)
-            else:
-                return data.reshape(*reshape_shape)
-        return data
+    @property
+    def n_rm_samples(self):
+        return self._n_rm_samples
+
+    def _reshape(self, data):
+        """
+        Reshape passed data, expected means or vars
+        :param data: flatten np.ndarray
+        :return: np.ndarray, reshaped data, the final data shape depends on the particular QType
+                             there is currently a reshape for ArrayType only
+        """
+        return self.qtype.reshape(data)
 
     def __getitem__(self, key):
         """
-        Get items from Quantity, quantity type must support brackets access
+        Get item from current QuantityMean, quantity type must support brackets access
+        All levels means and vars are reshaped to their QType shape and then the item is gotten,
+        ath the end, new QuantityMean instance is created with flatten selected means and vars
         :param key: str, int, tuple
-        :return: np.ndarray
+        :return: QuantityMean
         """
         new_qtype, start = self.qtype[key]  # New quantity type
         reshape_shape = None
-        newshape = None
+
         # ArrayType might be accessed directly regardless of qtype start and size
         if isinstance(self.qtype, qt.ArrayType):
```

Comment:
getitem should be common to the Quantity and thus implemented in QType as it is the part shared between Quantity and QuantityMean. It should go as:
```
new_qtype, slice = self.qtype.get_key(key)

samples_like = data[None, None,:]     # applied for data in [self._l_means, self._l_vars]
shaped_data = self.qtype.reshape(samples_like)     # actually reshape only for Array type
new_data = shaped_data[slice].ravel()
```